src/client/debugger/jupyter/kernelDebugAdapter.ts

```diff
@@ -0,0 +1,331 @@
+                        if (source && source.path) {
+                            const cell = this.fileToCell.get(source.path);
+                            if (cell) {
+                                source.name = path.basename(cell.document.uri.path);
+                                if (cell.index >= 0) {
+                                    source.name += `, Cell ${cell.index + 1}`;
+                                }
+                                source.path = cell.document.uri.toString();
+                            }
+                        }
+                    });
+
+                    this.sendMessage.fire(msg.content);
+                };
+                this.messageListener.set(message.seq, control);
+            }
+        } else if (message.type === 'response') {
+            // responses of reverse requests
+            const response = debugResponse(message as DebugProtocol.Response);
+            this.jupyterSession.requestDebug({
+                seq: response.content.seq,
+                type: 'request',
+                command: response.content.command
+            });
+        } else {
+            // cannot send via iopub, no way to handle events even if they existed
+            console.assert(false, `Unknown message type to send ${message.type}`);
+        }
+    }
+
+    dispose() {
+        this.messageListener.forEach((ml) => ml.dispose());
+        this.messageListener.clear();
+    }
+
+    private async dumpCellsThatRanBeforeDebuggingBegan() {
+        this.cellMap.getCellsAnClearQueue(this.notebookDocument).forEach(async (cell) => {
+            await this.dumpCell(cell.document.uri.toString());
+        });
+    }
+
+    /**
+     * Dump content of given cell into a tmp file and return path to file.
+     */
+    private async dumpCell(uri: string): Promise<void> {
+        const cell = this.notebookDocument.getCells().find((c) => c.document.uri.toString() === uri);
+        if (cell) {
+            try {
+                const response = await this.session.customRequest('dumpCell', { code: cell.document.getText() });
+                const norm = path.normalize(response.sourcePath);
+                this.fileToCell.set(norm, cell);
+                this.cellToFile.set(cell.document.uri.toString(), norm);
+            } catch (err) {
+                console.log(err);
+            }
+        }
+    }
+
+    private async debugInfo(): Promise<void> {
+        const response = await this.session.customRequest('debugInfo');
+
+        // If there's breakpoints at this point, delete them
+        if (response.breakpoints.lenght > 0) {
+            this.jupyterSession.requestDebug({
+                seq: 0,
+                type: 'request',
+                command: 'setBreakpoints',
+                arguments: {
+                    source: {
+                        path: response.breakpoints[0].source
```

Comment:
Will they all have the same source here?

Reply:
cleanup commit
Small change, but instead of deleting the breakpoints on the kernel, I instead sent a message to VS Code to register them.